File: vortex/src/core/cgid.js
"use strict";
/** The .cgid files NGIO writes, one per cell, and the seasonal names they are given afterwards. */
const fs = require("fs");
const path = require("path");
const { SEASONAL_SUFFIXES } = require("./seasons");

function isPlain(name) {
  const low = name.toLowerCase();
  return low.endsWith(".cgid") && !SEASONAL_SUFFIXES.some((s) => low.endsWith(`${s.toLowerCase()}.cgid`));
}

function list(dir) {
  try {
    return fs.readdirSync(dir).filter((n) => n.toLowerCase().endsWith(".cgid"));
  } catch {
    return [];
  }
}

/**
 * Move the cache files NGIO created this season (names not present in `before`) from the game's
 * Data/Grass into `dst`, adding the season suffix. Files that were already there belong to other
 * mods (Vortex hardlinks) and are never touched; so are new names `isForeign` says another mod deployed.
 * @param {(nameLower: string) => boolean} [isForeign]
 */
function moveNew(srcDir, before, dst, suffix, isForeign = () => false) {
  fs.mkdirSync(dst, { recursive: true });
  let moved = 0;
  for (const name of list(srcDir)) {
    if (before.has(name.toLowerCase()) || !isPlain(name) || isForeign(name.toLowerCase())) continue;
    const target = suffix ? name.slice(0, -".cgid".length) + suffix + ".cgid" : name;
    const from = path.join(srcDir, name);
    const to = path.join(dst, target);
    try {
      fs.renameSync(from, to);
    } catch (e) {
      if (e.code !== "EXDEV") throw e;          // staging on another drive: copy, then remove
      fs.copyFileSync(from, to);
      fs.unlinkSync(from);
    }
    moved += 1;
  }
  return moved;
}

/** DynDOLOD's grass LOD reads plain names: copy Name<suffix>.cgid to Name.cgid. */
function copyWithoutSuffix(srcDir, suffix, dstDir) {
  fs.mkdirSync(dstDir, { recursive: true });
  const tail = `${suffix}.cgid`.toLowerCase();
  let copied = 0;
  for (const name of list(srcDir)) {
    if (!name.toLowerCase().endsWith(tail)) continue;
    fs.copyFileSync(path.join(srcDir, name), path.join(dstDir, name.slice(0, -tail.length) + ".cgid"));
    copied += 1;
  }
  return copied;
}

module.exports = { isPlain, list, moveNew, copyWithoutSuffix };

File: mo2/ngio_grass/core/cgid.py
"""The .cgid files NGIO writes, one per cell, and the seasonal names they are given afterwards."""
from __future__ import annotations

import os
import shutil

from .seasons import SEASONAL_SUFFIXES


def is_plain(name: str) -> bool:
    """A cache file NGIO wrote itself, not yet given a season."""
    low = name.lower()
    return low.endswith(".cgid") and not any(low.endswith(s.lower() + ".cgid") for s in SEASONAL_SUFFIXES)


def count(grass_dir: str) -> int:
    if not os.path.isdir(grass_dir):
        return 0
    return sum(1 for n in os.listdir(grass_dir) if n.lower().endswith(".cgid"))


def add_suffix(grass_dir: str, suffix: str) -> int:
    """Rename every plain Name.cgid to Name<suffix>.cgid. Returns how many were renamed."""
    if not suffix or not os.path.isdir(grass_dir):
        return 0
    renamed = 0
    for name in os.listdir(grass_dir):
        if not is_plain(name):
            continue
        target = name[: -len(".cgid")] + suffix + ".cgid"
        os.replace(os.path.join(grass_dir, name), os.path.join(grass_dir, target))
        renamed += 1
    return renamed


def copy_without_suffix(src_dir: str, suffix: str, dst_dir: str) -> int:
    """DynDOLOD's grass LOD reads plain names: copy Name<suffix>.cgid to Name.cgid."""
    if not os.path.isdir(src_dir):
        return 0
    os.makedirs(dst_dir, exist_ok=True)
    tail = (suffix + ".cgid").lower()
    copied = 0
    for name in os.listdir(src_dir):
        if name.lower().endswith(tail):
            shutil.copy2(os.path.join(src_dir, name), os.path.join(dst_dir, name[: -len(tail)] + ".cgid"))
            copied += 1
    return copied
